services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -251,6 +258,170 @@ public URL getUrl(GetUrlRequest getUrlRequest) {
+     *
+     * <p>
+     * For more information on path-style and virtual-hosted-style URIs, see <a href=
+     * "https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html"
+     * >Methods for accessing a bucket</a>.
+     *
+     * @param uri The URI to be parsed
+     * @return Parsed {@link S3Uri}
+     *
+     * <p><b>Example Usage</b>
+     * <p>
+     * {@snippet :
+     *     S3Client s3Client = S3Client.create();
+     *     S3Utilities s3Utilities = s3Client.utilities();
+     *     String uriString = "https://myBucket.s3.us-west-1.amazonaws.com/doc.txt?versionId=abc123";
+     *     URI uri = URI.create(uriString);
+     *     S3Uri s3Uri = s3Utilities.parseUri(uri);
+     *
+     *     String bucket = s3Uri.bucket().orElse(null); // "myBucket"
+     *     String key = s3Uri.key().orElse(null); // "doc.txt"
+     *     Region region = s3Uri.region().orElse(null); // Region.US_WEST_1
+     *     boolean isPathStyle = s3Uri.isPathStyle(); // false
+     *     String versionId = s3Uri.firstMatchingRawQueryParameter("versionId").orElse(null); // "abc123"
+     *}
+     */
+    public S3Uri parseUri(URI uri) {
+        validateUri(uri);
+
+        if ("s3".equalsIgnoreCase(uri.getScheme())) {
+            return parseAwsCliStyleUri(uri);
+        }
+
+        return parseStandardUri(uri);
+    }
+
+    private S3Uri parseStandardUri(URI uri) {
+        String bucket;
+        String key;
+        String region = null;
+        boolean isPathStyle = false;
+        Map<String, List<String>> queryParams = new HashMap<>();
+
+        if (uri.getHost() == null) {
+            throw new IllegalArgumentException("Invalid S3 URI: no hostname: " + uri);
+        }
+
+        Matcher matcher = ENDPOINT_PATTERN.matcher(uri.getHost());
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Invalid S3 URI: hostname does not appear to be a valid S3 endpoint: " + uri);
+        }
+
+        String[] parsed;
+        String prefix = matcher.group(1);
+        if (prefix == null || prefix.isEmpty()) {
+            isPathStyle = true;
+            parsed = parsePathStyleUri(uri);
+        } else {
+            parsed = parseVirtualHostedStyleUri(uri, matcher);
+        }
+        key = parsed[0];
+        bucket = parsed[1];
+
+        if (!"amazonaws".equals(matcher.group(2))) {
+            region = matcher.group(2);
+        }
+
+        String queryPart = uri.getQuery();
+        if (queryPart != null) {
+            queryParams = SdkHttpUtils.uriParams(uri);
+        }
+
+        Region uriRegion = region != null ? Region.of(region) : null;
+
+        return S3Uri.builder()
+                    .uri(uri)
+                    .bucket(bucket)
+                    .key(key)
+                    .region(uriRegion)
+                    .isPathStyle(isPathStyle)
+                    .queryParams(queryParams)
+                    .build();
+
+    }
+
+    private String[] parsePathStyleUri(URI uri) {
```

Comment:
I'd use `Pair<String, String>` instead of `String[]`

Reply:
Pair throws an exception when either the key is bucket is null